yjit_codegen.c

```diff
@@ -3405,9 +3389,32 @@ gen_send_cfunc(jitstate_t *jit, ctx_t *ctx, const struct rb_callinfo *ci, const
         // The method gets a pointer to the first argument
         // rb_f_puts(int argc, VALUE *argv, VALUE recv)
         mov(cb, C_ARG_REGS[0], imm_opnd(argc));
-        lea(cb, C_ARG_REGS[1], mem_opnd(64, RAX, -(argc) * SIZEOF_VALUE));
-        mov(cb, C_ARG_REGS[2], mem_opnd(64, RAX, -(argc + 1) * SIZEOF_VALUE));
+        lea(cb, C_ARG_REGS[1], ctx_stack_opnd(ctx, argc - 1));
+        mov(cb, C_ARG_REGS[2], ctx_stack_opnd(ctx, argc));
     }
+    // Variadic method with Ruby array
+    if (cfunc->argc == -2) {
+        // Create an array from the arguments.
+        // This will allocate, which is safe because our SP is already saved in
+        // the future-callee's CFP.
```

Comment:
Thanks for checking that it's safe. Could you comment that safety follows from having the same ordering of events as `vm_call_cfunc_with_frame()`? We push the frame by changing `ec->cfp` and then we do the allocation. I would link to `call_cfunc_invoker_func()` and `call_cfunc_m2()` and note that we use `rb_ec_ary_new_from_values()` instead of `rb_ary_new4()` to do the allocation.

I also don't like the word "safe" because beyond safety we also need to keep the order consistent as in case of `NoMemoryError`, the stack trace is user observable.